lib/experimental/html/class-wp-html-attribute-sourcer.php

```diff
@@ -0,0 +1,394 @@
+
+		while ( $at < strlen( $s ) && ',' !== $s[ $at ] ) {
+			/*
+			 * Descendant combinators are harder to discover because we
+			 * always have to skip whitespace, but that whitespace could
+			 * be the combinator if we don't approach anything else first.
+			 */
+			$ws_length = strspn( $s, " \f\n\r\t", $at );
+			$at += $ws_length;
+
+			if ( !$is_first && $ws_length > 0 && 0 === strspn( $s[ $at ], '>+~' ) ) {
+				$at--;
+				$s[ $at ] = ' ';
+			}
+			$is_first = false;
+
+			switch ( $s[ $at ] ) {
+				case '>':
+				case '+':
+				case '~':
+				case ' ':
+					$combinator = $s[ $at ];
+					$at++;
+					$at += strspn( $s, " \f\n\r\t", $at );
+					$inner = self::parse_selector( $s, $at );
+					if ( null === $inner ) {
+						return null;
+					}
+					list( $inner_selector, $next_at ) = $inner;
+					$inner_selector['combinator'] = $combinator;
+					$selector['then'] = $inner_selector;
+					$at = $next_at;
+					break;
+
+				case '.':
+					$at++;
+					$class_name = self::parse_css_identifier( $s, $at );
+					if ( null === $class_name ) {
+						return null;
+					}
+
+					if ( ! isset( $selector['class_names'] ) ) {
+						$selector['class_names'] = array();
+					}
+					$selector['class_names'][] = $class_name;
+					$at += strlen( $class_name );
+					break;
+
+				case '#':
+					$at++;
+					// @TODO: Hashes don't have to start with `nmstart` so this might reject valid hash names.
+					$element_id = self::parse_css_identifier( $s, $at );
+					if ( null === $element_id ) {
+						return null;
+					}
+
+					$selector['hash'] = $element_id;
+					$at += strlen( $element_id );
+					break;
+
+				case '[':
+					/*
+					 * Only current support is for checking of presence of attributes
+					 * with a very-limited subset of allowable names, not whether the
+					 * attribute conforms to a given value or is allowed in HTML.
+					 */
+					$at++;
+					$inside_length = strspn( $s, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-", $at );
+					if ( ']' !== $s[ $at + $inside_length ] ) {
+						return null;
+					}
+
+					$attribute = substr( $s, $at, $inside_length );
+					$at += $inside_length + 1;
+
+					$selector['has_attribute'] = $attribute;
+					break;
+
+				default:
+					$tag_name = self::parse_css_identifier( $s, $at );
+					if ( null === $tag_name ) {
+						return null;
+					}
+
+					$selector['tag_name'] = $tag_name;
+					$at += strlen( $tag_name );
+			}
+		}
+
+		return [ $selector, $at ];
+	}
+
+	/**
+	 * Parses CSS identifier; currently limited to ASCII identifiers.
+	 *
+	 * Example:
+	 * ```
+	 *     'div' === parse_css_identifier( 'div > img' );
+	 * ```
+	 *
+	 * Grammar:
+	 * ```
+	 *     ident        -?{nmstart}{nmchar}*
+	 *     nmstart      [_a-z]|{nonascii}|{escape}
+	 *     nmchar       [_a-z0-9-]|{nonascii}|{escape}
+	 *     nonascii     [\240-\377]
+	 *     escape       {unicode}|\\[^\r\n\f0-9a-f]
+	 *     unicode      \\{h}{1,6}(\r\n|[ \t\r\n\f])?
+	 *     h            [0-9a-f]
```

Comment:
h -> hex? For readability – I got confused for a moment

Reply:
this is just a direct copy from the spec. yes, `h` is there for a hex digit